El 3593 icon documentation

TaraTodd123 opened the pull request:
#### Checklist

* [x] The contents of this PR are mine to submit. No third party code or other material is being committed.
* [x] New third party dependencies (if any) are linked via `package.json`. Third party dependencies are licenced under one of the following: Apache, MIT, BSD, Mozilla Public License, Eclipse Public License, or Oracle Binary Code License.
* [x] The code in this PR does not contain export-restricted encryption algorithms.

#### Ticket / Issue

https://portal.digitalsafe.net/browse/EL-3593

#### Description of Proposed Changes

- Separate ux-icon component documentation added
- list of icons on shows name for ux-icon component
- when clicking on list of icons a code snippet of its implementation is copied

#### Documentation CI URL

https://pages.github.houston.softwaregrp.net/sepg-docs-qa/UXAspects_CI_UXAspects_EL-3593-Icon-Documentation

ashley-hunter commented on docs/app/data/css-page.json:
```diff
@@ -264,6 +264,21 @@
                     "component": "CssUxIconsComponent"
                 },
+                {
+                    "title": "Icons Documentation",
```
Might want to reconsider this title, every section is "documentation" so this seems a bit redundant. Also, as this section is an Angular component should this be in the components section rather than CSS? thoughts @donna-marijne?

donna-marijne replied:
Agreed, we should establish an "Icons" category under "Components," and the section should be named "Icon." The documentation section containing the list of icons should be renamed to "Icon Set."

TaraTodd123 replied:
Moved to an Icons Section under components and renamed list of icons to Icon Set

ashley-hunter commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.ts:
```diff
@@ -0,0 +1,63 @@
+    public copied: boolean = false;
+
+    hideTriggers: string [] = ['mouseleave', 'blur'];
+    showTriggers: string [] = ['focus', 'mouseenter'];
```
This can be removed, don't think it is used.

TaraTodd123 replied:
Removed

ashley-hunter commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.ts:
```diff
@@ -0,0 +1,63 @@
+    @Input() iconset: string;
+    @Input() iconClass: string;
+
+    public tooltipText: string = 'Click to copy icon text';
```
Could we remove the `public` keyword, by default properties are public so we normally just omit this.

TaraTodd123 replied:
Removed

ashley-hunter commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.ts:
```diff
@@ -0,0 +1,63 @@
+
+        switch (this.iconset) {
+            case 'ux-icons':
+                dummy.value =  '<i class="ux-icon' + ' ' + this.classname + '"></i>';
```
There seems to be a double space after the equal sign, there also seems to be an unnecessary space: `+ ' ' +`. We can use backticks to make this easier: 

```typescript
dummy.value = `<i class="ux-icon ${this.classname}"></i>`;
```

TaraTodd123 replied:
changed to use <code>${this.classname}</code>

ashley-hunter commented on docs/app/pages/css/css-sections/icons/ux-icons/ux-icons.component.ts:
```diff
@@ -1,21 +1,24 @@
 @DocumentationSectionComponent('CssUxIconsComponent')
 export class CssUxIconsComponent {
 
+    public copied: boolean = false;
```
Is this used?

TaraTodd123 replied:
Removed

ashley-hunter commented on docs/app/pages/css/css-sections/icons/icon-documentation/snippets/service.ts:
```diff
@@ -0,0 +1,17 @@
+    constructor(iconService: IconService) {
+
+        // replace single icon
+        iconService.setIcon({ name: '3d', icon: 'ux-icon-3d', iconset: 'ux-icons', size: '14px' });
```
Perhaps these examples should use a different iconset, eg, quantum icons (http://quantum-ux.swinfra.net:8011/#!/components/icons) as currently this isn't really overriding anything?

TaraTodd123 replied:
Changed to use quantum icons in examples

ashley-hunter commented on docs/app/pages/css/css-sections/icons/icon-documentation/icon-documentation.component.html:
```diff
@@ -0,0 +1,67 @@
+<p>Add this code snippet to your app.module.ts file to override icons within your whole application from a different icon set, by specifying the <code>name</code> of the icon you want to replace, the <code>icon</code> you want to replace it with and the <code>iconset</code> you want to replace it from. Optionally you can add a <code>size</code> parameter.</p>
+<uxd-snippet [content]="snippets.compiled.appModuleTs" language="javascript"></uxd-snippet>
+
+<h4>Single Component</h4>
```
This would be overriding on a per module basis rather than a single component basis.

TaraTodd123 replied:
Changed

donna-marijne commented on docs/app/pages/css/css-sections/icons/icon-documentation/icon-documentation.component.html:
```diff
@@ -0,0 +1,67 @@
+
+<h3>Overriding Custom Iconset</h3>
+
+<h4>Whole Application</h4>
```
I'd suggest combining this with the following section, and adding a comment that `forRoot` should be replaced with `forChild` if the module is not the root module.

TaraTodd123 replied:
Updated to include the above

donna-marijne commented on docs/app/pages/css/css-sections/icons/icon-documentation/icon-documentation.component.html:
```diff
@@ -0,0 +1,67 @@
+<uxd-snippet [content]="snippets.compiled.singleModuleTs" language="javascript"></uxd-snippet>
+
+<h4>Single Component(Alternative)</h4>
+<p>An alternative way is to use the iconService add this code snippet to your module file to override icons within the specific module, by specifying the <code>name</code> of the icon you want to replace, the <code>icon</code> you want to replace it with and the <code>iconset</code> you want to replace it from. Optionally you can add a <code>size</code> parameter.</p>
```
Could you document the functions available on the `IconService` using the standard API table?  

Also the service is named IconService (note capitalization), and since it is a type we would highlight it with `<code>`.

TaraTodd123 replied:
Documented <code>IconService</code>

TaraTodd123 commented on docs/app/pages/css/css-sections/icons/icon-documentation/icon-documentation.component.html:
```diff
@@ -0,0 +1,67 @@
+
+<hr>
+
+<p><code>ux-icon</code> can be used to allow a user to dynamically add icons.</p>
```
Updated to include the above

TaraTodd123 commented on docs/app/pages/css/css-sections/icons/icon-documentation/icon-documentation.component.html:
```diff
@@ -0,0 +1,67 @@
+
+<uxd-api-properties tableTitle="Inputs">
+    <tr uxd-api-property name="name" type="string">
+        Defines the icon to be displayed.
```
Updated to include the above

TaraTodd123 commented on docs/app/pages/css/css-sections/icons/icon-documentation/icon-documentation.component.html:
```diff
@@ -0,0 +1,67 @@
+        Defines the icon to be displayed.
+    </tr>
+    <tr uxd-api-property name="size" type="string">
+        Defines a custom size for the icon.
```
Updated to include the above

TaraTodd123 commented on docs/app/pages/css/css-sections/icons/icon-documentation/icon-documentation.component.html:
```diff
@@ -0,0 +1,67 @@
+        Defines a custom size for the icon.
+    </tr>
+    <tr uxd-api-property name="rotate" type="number">
+        Defines the number of degrees to rotate the icon either <code>90</code>, <code>180</code>, <code>270</code>.
```
Updated to include the above

TaraTodd123 commented on docs/app/pages/css/css-sections/icons/icon-documentation/icon-documentation.component.html:
```diff
@@ -0,0 +1,67 @@
+    </tr>
+</uxd-api-properties>
+
+<h3>Overriding Custom Iconset</h3>
```
Updated to the above

TaraTodd123 commented on docs/app/pages/css/css-sections/icons/icon-documentation/icon-documentation.component.html:
```diff
@@ -0,0 +1,67 @@
+<p>Add this code snippet to a components module file to override icons within a specific module, by specifying the <code>name</code> of the icon you want to replace, the <code>icon</code> you want to replace it with and the <code>iconset</code> you want to replace it from. Optionally you can add a <code>size</code> parameter.</p>
+<uxd-snippet [content]="snippets.compiled.singleModuleTs" language="javascript"></uxd-snippet>
+
+<h4>Single Component(Alternative)</h4>
```
Updated to previous comment

donna-marijne commented on docs/app/pages/components/components-sections/icon/icon/icon.component.html:
```diff
@@ -0,0 +1,124 @@
+
+<uxd-api-properties tableTitle="Icon Service">
+    <tr uxd-api-property name="setIcon" args="IconDefinition">
+        This function should be called to override a single icon from the default icon set. It should be
```
Note that it's not just for overriding; one can define an entirely new icon with an arbitrary name. I would suggest summarizing this as "Create a single icon definition. See the `IconDefinition` interface below for details."

donna-marijne commented on docs/app/pages/components/components-sections/icon/icon/icon.component.html:
```diff
@@ -0,0 +1,124 @@
+    </tr>
+</uxd-api-properties>
+
+<p>The object to be passed to the <code>setIcon</code> and <code>setIcons</code> functions. The following
```
Add the "IconDefinition Interface" title above this and remove `tableTitle` below.

This interface is also used in the `forRoot`/`forChild` configuration. I would summarize this as "This interface defines the properties used to create a custom icon definition."

donna-marijne commented on docs/app/pages/components/components-sections/icon/icon/icon.component.html:
```diff
@@ -0,0 +1,124 @@
+</uxd-api-properties>
+
+<p>The object to be passed to the <code>setIcon</code> and <code>setIcons</code> functions. The following
+    options can be passed in this object to configure the overriding of the default icon set:</p>
```
As noted above, this is not limited to overriding the default icon set. I would drop this sentence and start a new paragraph to say that custom icon definitions will take precendence over the default icon set, and can therefore be used to customize the icons that appear within UX Aspects components.

donna-marijne commented on docs/app/pages/components/components-sections/icon/icon/icon.component.html:
```diff
@@ -0,0 +1,124 @@
+
+<uxd-api-properties tableTitle="IconDefinition Interface">
+    <tr uxd-api-property name="name" type="string">
+        The name of the icon to be overridden in the default icon set.
```
Suggestion: "The name of the icon. This corresponds to the `name` property of `ux-icon`."

donna-marijne commented on docs/app/pages/components/components-sections/icon/icon/icon.component.html:
```diff
@@ -0,0 +1,124 @@
+        The name of the icon to be overridden in the default icon set.
+    </tr>
+    <tr uxd-api-property name="icon" type="string">
+        The name of the icon to override the default icon.
```
Should be something like: "The CSS class of the font icon which should be rendered for the above `name`, e.g. `qtm-icon-action`."

donna-marijne commented on docs/app/pages/components/components-sections/icon/icon/icon.component.html:
```diff
@@ -0,0 +1,124 @@
+        The name of the icon to override the default icon.
+    </tr>
+    <tr uxd-api-property name="iconset" type="string">
+        The name of the icon set that will be used to override the default icon set.
```
Similar to above, include an example to make this clear, e.g. "The CSS class of the font icon set containing the given `icon`, e.g. `qtm-font-icon`."

donna-marijne commented on docs/app/pages/components/components-sections/icon/icon/icon.component.html:
```diff
@@ -0,0 +1,124 @@
+        The name of the icon set that will be used to override the default icon set.
+    </tr>
+    <tr uxd-api-property name="size" type="string">
+        An optional parameter that can be used to override icons to an exact size.
```
I believe this can be either a single string or an array, so the `type` property should be updated.

Could you move this row to appear below `name` since it is part of the selector to match a specific icon?

Instead of noting that it is option, please specify `required="true"` on all the other rows.

Suggested text: "The size(s) at which this icon is defined, in CSS units. This corresponds to the `size` property of `ux-icon`. This property can be used to provide an alternate icon glyph at different sizes, as an alternative to the browser scaling which would otherwise apply. Note that due to technical limitations this only applies to icons explicitly sized using the `size` property, and cannot match a size defined in CSS."

donna-marijne commented on docs/app/pages/components/components-sections/icon/icon/icon.component.html:
```diff
@@ -0,0 +1,124 @@
+
+<uxd-snippet [content]="snippets.compiled.appHtml" language="html"></uxd-snippet>
+
+<h4>Overriding the Default Icon Set</h4>
```
On reflection, this should probably be "Defining Custom Icons."

We should also add a paragraph describing why one would want to do this. The use cases are:
1. Define new icons which can be rendered using `ux-icon`.
2. Override the default icons which appear within UX Aspects components.
3. Provide an alternate icon glyph to display when an icon is rendered at a smaller or larger size.

donna-marijne commented on docs/app/pages/components/components-sections/icon/icon/icon.component.html:
```diff
@@ -0,0 +1,124 @@
+
+<h4>Overriding the Default Icon Set</h4>
+
+<p>Add this code snippet to your root module file to override icons within your whole application from a
```
Instead of "override" use "define."

TaraTodd123 commented:
Thanks for the feedback, changed icon set styling to as stated above with the popover and clipboard icon. Also updated the <code>ux-icon</code> documentation as stated above.

ashley-hunter commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.ts:
```diff
@@ -0,0 +1,40 @@
+    @Input() iconset: string;
+    @Input() iconClass: string;
+
+    @ViewChild(TooltipDirective, { static: true }) tooltip: TooltipDirective;
```
I don't think this is ever used

ashley-hunter commented on docs/styles.less:
```diff
@@ -1598,4 +1598,15 @@ Spark table documentation styles
     z-index: 1;
+}
+
+.ux-icon-component-example ux-icon {
```
We only put global styles in this stylesheet. These styles only affect a single documentation section so should go in the component stylesheet.

Also any styles you use in the documentation will also need to be added to the plunker.

Also the `nth-of-type(3)` selector is very rigid, if we ever wanted to add additional icons to the example the styling could then be wrong. A better way would be to have a container div element that groups the icons and adds `margin-right`.

ashley-hunter commented on docs/app/pages/components/components-sections/icon/icon/icon.component.html:
```diff
@@ -0,0 +1,136 @@
+        Rotation of the icon in degrees.
+    </tr>
+    <tr uxd-api-property name="flipHorizontal" type="boolean" defaultValue="false">
+        The icon can be horizontally flipped if set to true. The default is false.
```
Could we put boolean values inside `<code>` elements as they are keywords

ashley-hunter commented on docs/app/pages/components/components-sections/icon/icon/icon.component.html:
```diff
@@ -0,0 +1,139 @@
+<div class="ux-icon-component-example">
```
I think we can simplify this a little bit: https://plnkr.co/edit/ZJcHLEYlgp42n4lN9exY?p=preview
Its also good to avoid the `ux` prefix in sample code classes as we sometimes use this to essentially namespace our styles to avoid class collisions.

TaraTodd123 replied:
Updated to plunker example

ashley-hunter commented on docs/app/pages/components/components-sections/icon/icon/snippets/appModule.ts:
```diff
@@ -0,0 +1,17 @@
+        })
+    ],
+})
+
```
Can we remove this empty line between the decorator and class

TaraTodd123 replied:
Removed

ashley-hunter commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.html:
```diff
@@ -0,0 +1,42 @@
+                    (click)="copy('ux-icon-component')"></i></td>
+        </tr>
+        <tr *ngIf="iconset === 'ux-icons'">
+            <td class="p-l-xs">&lt;i class="ux-icon {{classname}}"&gt;&lt;/i&gt;&nbsp;&nbsp;</td>
```
I think the code snippets in the popover would look better if we used the code highlighting like we do elsewhere

TaraTodd123 replied:
Used the code highlighting for popover

ashley-hunter commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.html:
```diff
@@ -0,0 +1,42 @@
+        </tr>
+        <tr *ngIf="iconset !== 'ux-icons'">
+            <td class="p-l-xs">&lt;i class="hpe-icon {{classname}}"&gt;&lt;/i&gt;&nbsp;&nbsp;</td>
+            <td><i class="ux-icon ux-icon-clipboard m-r-xs"
```
Could we also improve the alignment and indentation of some of these tags, eg:

```html
<td><i class="ux-icon ux-icon-clipboard m-r-xs"
                   [uxTooltip]="tooltipTemplate"
                    (click)="copy('hpe-icon')"></i></td>
```
Should be:
```html
<td
    <i class="ux-icon ux-icon-clipboard m-r-xs"
       [uxTooltip]="tooltipTemplate"
       (click)="copy('hpe-icon')"
    </i>
</td>
```

TaraTodd123 replied:
improved alignment

ashley-hunter commented on docs/styles.less:
```diff
@@ -1598,4 +1598,8 @@ Spark table documentation styles
     z-index: 1;
+}
+
+.icon-code-snippets {
```
Could we possibly alter this class name to indicate that this class is for the popover as I initially assumed this was for the code snippets that displays the code needed to use an icon

TaraTodd123 replied:
renamed

ashley-hunter commented on docs/app/pages/css/css-sections/icons/ux-icons/ux-icons.component.html:
```diff
@@ -21,9 +24,12 @@
-            <i class="{{ iconClass }} p-r-xs" [ngClass]="icon.classname"></i> <span>{{ icon.classname }}</span>
-        </div>
+    <div class="col-md-4 col-sm-4 icon-demo-col m-b" *ngFor="let icon of icons; trackBy: trackByFn">
+        <uxd-icon-preview
```
I wonder would it be better if the buttons were uniform sizes? They currently are like this:

[![Image from Gyazo](https://i.gyazo.com/ab20bbc7ca89114704f5db7281ce1f00.gif)](https://gyazo.com/ab20bbc7ca89114704f5db7281ce1f00)

Perhaps this would be better?

[![Image from Gyazo](https://i.gyazo.com/d7061e72bc1047cffb727bcc12ac9c16.gif)](https://gyazo.com/d7061e72bc1047cffb727bcc12ac9c16)

@donna-marijne thoughts?

donna-marijne replied:
Agree, consistent hit box will be easier to use.

TaraTodd123 replied:
hover now full width

ashley-hunter commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.html:
```diff
@@ -0,0 +1,42 @@
+</ng-template>
+
+<ng-template #tooltipTemplate>
+    <p class="usage-tooltip">Copy to Clipboard</p>
```
Looks like this `p` tag has margin on the bottom making the popover have empty space below:

![image](https://user-images.githubusercontent.com/20795331/62369946-1d37e600-b529-11e9-8983-c1f97caaaa4a.png)

donna-marijne replied:
I'd suggest only using `p` for actual paragraphs of text, otherwise a div is more appropriate.

TaraTodd123 replied:
changed to <code>div</code> rather than <code>p</code>

ashley-hunter commented on docs/app/pages/components/components-sections/icon/icon/icon.component.html:
```diff
@@ -0,0 +1,139 @@
+
+<p>The following code can be used to create the example above:</p>
+
+<uxd-snippet [content]="snippets.compiled.appHtml" language="html"></uxd-snippet>
```
We require some CSS to create the example above, we should probably document it here too, otherwise someone copying this code would not get exactly the same result as shown above

TaraTodd123 replied:
Updated to have tabs for code snippet

donna-marijne commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.html:
```diff
@@ -0,0 +1,64 @@
+                <td>
+                    <i class="ux-icon ux-icon-clipboard m-r-xs"
+                       [uxTooltip]="tooltipTemplate"
+                       (click)="copy('ux-icon-component')">
```
If something is a button, then it should be a `button`. We have styles for a [borderless icon-only button](https://pages.github.houston.softwaregrp.net/caf/ux-aspects-micro-focus/#/ui-components/buttons#icon-buttons-documentation-component) with proper mouseover and focus support, so let's leverage that here.

TaraTodd123 replied:
changed

ashley-hunter commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.html:
```diff
@@ -0,0 +1,64 @@
+<button uxFocusIndicator
```
I'm noticing some strange hover behaviour. The hover background colour now stretches the full width which is as expected, however hovering the mouse over the area does not show the hover effect until the mouse is near the text:

[![Image from Gyazo](https://i.gyazo.com/862c829911bc5bcc558a96bfe6c95dd8.gif)](https://gyazo.com/862c829911bc5bcc558a96bfe6c95dd8)

TaraTodd123 replied:
Mouse over area is now full width of button

ashley-hunter commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.html:
```diff
@@ -0,0 +1,64 @@
+            </tr>
+            <tr *ngIf="iconset === 'ux-icons'">
+                <td class="p-l-xs">
+                    <code>
```
Would it be possible to try and use this style of code snippet? We normally use this style for actual code and the `<code>` element for keywords or classes etc.. rather than actual full lines of code

![image](https://user-images.githubusercontent.com/20795331/62453803-01bc1d80-b76b-11e9-97f6-c1441e44ad45.png)

TaraTodd123 replied:
Changed style of code snippet

ashley-hunter commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.html:
```diff
@@ -0,0 +1,64 @@
+<button uxFocusIndicator
+        tabindex="0"
+        class="icon-demo btn btn-link button-secondary btn-text"
+        [uxPopover]="iconCodeSnippet"
```
When I try to select the copy button the popover jumps (this may be related to the issue above):

[![Image from Gyazo](https://i.gyazo.com/ffb1d169bcf2f5587e8ff0d06ef51299.gif)](https://gyazo.com/ffb1d169bcf2f5587e8ff0d06ef51299)

TaraTodd123 replied:
Fixed with the above fix

donna-marijne commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.html:
```diff
@@ -0,0 +1,64 @@
+                    </i>
+                </td>
+            </tr>
+            <tr *ngIf="iconset !== 'ux-icons'">
```
Condition should be `iconset === 'hpe-icons'` since you have explicitly cited `hpe-icon` below.

TaraTodd123 replied:
updated to above

ashley-hunter commented on docs/app/pages/components/components-sections/icon/icon/icon.component.html:
```diff
@@ -0,0 +1,145 @@
+        Rotation of the icon in degrees.
+    </tr>
+    <tr uxd-api-property name="flipHorizontal" type="boolean" defaultValue="false">
+        The icon can be horizontally flipped if set to <code>true</code>. The default is <code>false</code>.
```
I think we can remove `The default is false` from the paragraph as the default value has been specified in the attribute. Same for the `flipVertical` input.

TaraTodd123 replied:
Removed

donna-marijne commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.less:
```diff
@@ -0,0 +1,36 @@
+@import '../../../../../../assets/styles/variables.less';
+
+.btn-text {
```
Could this be something like `.uxd-icon-button` or something? Bit more descriptive. `uxd` is the prefix that we use for stuff that only applies to the documentation, so it won't collide with anything else.

TaraTodd123 replied:
updated

donna-marijne commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.less:
```diff
@@ -0,0 +1,36 @@
+
+.button-secondary.btn-link:hover,
+.btn-white.btn-link:hover {
+    width: 100%;
```
Don't change the dimension on hover. Width should be 100% at all times for a consistent mouseover target. Also, you established a unique class for the icon buttons so please use that in the selector.

TaraTodd123 replied:
Updated to whole element rather than hover

ashley-hunter commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.ts:
```diff
@@ -0,0 +1,38 @@
+
+        dummy.select();
+        document.execCommand('copy');
+        this._renderer.removeChild(this._elementRef.nativeElement, dummy);
```
I think we avoided this as it doesn't work in IE

donna-marijne replied:
Makes sense, thank you for the clarification.

TaraTodd123 commented on docs/app/pages/components/components-sections/icon/icon/icon.component.html:
```diff
@@ -0,0 +1,145 @@
+
+<h4>Defining Custom Icons</h4>
+
+<p>Custom icons may want to be implemented to define new icons which can be rendered using
```
Changed

TaraTodd123 commented on docs/app/pages/components/components-sections/icon/icon/icon.component.html:
```diff
@@ -0,0 +1,145 @@
+<h4>Defining Custom Icons</h4>
+
+<p>Custom icons may want to be implemented to define new icons which can be rendered using
+    <code>ux-icon</code>, to override the default icons which appear within UX Aspects components or to provide
```
Changed

TaraTodd123 commented on docs/app/pages/components/components-sections/icon/icon/icon.component.html:
```diff
@@ -0,0 +1,145 @@
+    <tr uxd-api-property name="setIcons" args="Array<IconDefinition>">
+        This function should be called to override several icons from the default icon set in an array. It
+        should be passed an array of <code>IconDefinition</code> objects containing properties to override
+        the default icon set.
```
Updated

TaraTodd123 commented on docs/app/pages/components/components-sections/icon/icon/icon.component.html:
```diff
@@ -0,0 +1,145 @@
+    <tr uxd-api-property name="name" type="string" required="true">
+        The name of the icon. This corresponds to the <code>name</code> property of <code>ux-icon</code>.
+    </tr>
+    <tr uxd-api-property name="size" type="string | array">
```
Updated

TaraTodd123 commented on docs/app/pages/css/css-sections/icons/ux-icons/ux-icons.component.html:
```diff
@@ -1,13 +1,16 @@
-<p>To use an icon simply apply the <code>{{ iconClass }}</code> class along with one of the following classes:</p>
+<p>To use an icon simply apply the <code>{{ iconClass }}</code> class along with one of the following
+    classes <ng-container *ngIf="iconset === 'ux-icons'">or use the <code>ux-icon</code><a
```
Updated for each case

TaraTodd123 commented on docs/app/pages/components/components-sections/icon/icon/icon.component.html:
```diff
@@ -0,0 +1,145 @@
+<hr>
+
+<p>The <code>ux-icon</code> component can be used to dynamically display icons and allows easy control of
+    their size and orientation. Icons can be selected from the UX Aspects <a [routerLink]="['/css/icons']">
```
Added in re-direct

TaraTodd123 commented on docs/app/pages/css/css-sections/icons/ux-icons/ux-icons.component.html:
```diff
@@ -1,13 +1,16 @@
+<p>To use an icon simply apply the <code>{{ iconClass }}</code> class along with one of the following
+    classes <ng-container *ngIf="iconset === 'ux-icons'">or use the <code>ux-icon</code><a
+    [routerLink]="['/components/icons']"> component</a> along with one of the following
+    names</ng-container>. Click on an icon to view its implementation.</p>
```
Added in re-direct

TaraTodd123 commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.ts:
```diff
@@ -0,0 +1,38 @@
+    @Input() name: string;
+    @Input() classname: string;
+    @Input() iconset: string;
+    @Input() iconClass: string;
```
Changed

TaraTodd123 commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.html:
```diff
@@ -0,0 +1,64 @@
+                <td class="p-l-xs">
+                    <code>
+                        &lt;ux-icon name="{{name}}"&gt;&lt;/ux-icon&gt;
+                    </code>&nbsp;&nbsp;
```
Updated to its own component so it is more reusable

TaraTodd123 commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.html:
```diff
@@ -0,0 +1,64 @@
+                    </code>&nbsp;&nbsp;
+                </td>
+                <td>
+                    <i class="ux-icon ux-icon-clipboard m-r-xs"
```
Changed

ashley-hunter commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-snippet/icon-snippet.component.less:
```diff
@@ -0,0 +1,20 @@
+    }
+}
+
+::ng-deep pre[class*="language-"] {
```
This should be moved inside `:host` as ng-deep adds no limitations on the style, this means every `pre` tag on the site with a language class will receive this styling, not just the ones in the component.

TaraTodd123 replied:
moved inside <code>:host</code>

ashley-hunter commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-snippet/icon-snippet.component.less:
```diff
@@ -0,0 +1,20 @@
+:host {
+   display: inline-flex;
```
Indentation is a bit off here

TaraTodd123 replied:
fixed

ashley-hunter commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-snippet/icon-snippet.component.less:
```diff
@@ -0,0 +1,20 @@
+    width: 100%;
+
+    uxd-snippet {
+      display:flex;
```
Space missing between property name and value

TaraTodd123 replied:
space added

ashley-hunter commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-snippet/icon-snippet.component.less:
```diff
@@ -0,0 +1,20 @@
+        flex: 1;
+    }
+
+    button {
```
Might be good to use classes here instead of element selectors in case we ever add another button in to this component then it may get these styles unexpectedly

TaraTodd123 replied:
Added classes for elements

ashley-hunter commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-snippet/icon-snippet.component.ts:
```diff
@@ -0,0 +1,49 @@
+    styleUrls: ['./icon-snippet.component.less'],
+    changeDetection: ChangeDetectionStrategy.OnPush
+})
+
```
Remove space between decorator and class.

TaraTodd123 replied:
Removed space

ashley-hunter commented on docs/app/pages/components/components-sections/icon/icon/icon.component.ts:
```diff
@@ -1,14 +1,18 @@
-import { IPlayground } from '../../../../../interfaces/IPlayground';
-import { IPlaygroundProvider } from '../../../../../interfaces/IPlaygroundProvider';
-import { DOCUMENTATION_TOKEN, DocumentationType } from '../../../../../services/playground/tokens/documentation.token';
+import {ChangeDetectionStrategy, Component, Inject} from '@angular/core';
```
Could you check your editor settings, perhaps you have something like `prettier` enabled, it seems to be altering spaces in imports, we normally have a space after the bracket: `import { ChangeDetectionStrategy`, otherwise every file you edit from now on will add a lot of unnecessary formatting changes

donna-marijne commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-snippet/icon-snippet.component.html:
```diff
@@ -0,0 +1,15 @@
+        type="button"
+        class="btn btn-link btn-icon button-secondary snippet-copy-button"
+        aria-label="Copy to clipboard">
+    <ux-icon name="clipboard" aria-hidden="true"></ux-icon>
```
Remove aria-hidden, I don't think it's needed for the icon since it has no text. If it turns out that it is needed then we can add aria-hidden directly to the ux-icon component itself.

donna-marijne commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.ts:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    getUXIconSnippet(): string {
+        return `<i class="ux-icon ${this.classname}"></i>`;
```
Can't we use `iconSetClass` instead of hard coding `ux-icon` here, therefore eliminating the need for separate functions?

Also, I don't believe that the inputs ever change on this component (correct me if I am wrong) so you can generate the snippet strings once within `ngOnInit` and provide them to the view as properties.

TaraTodd123 commented:
Fixed issues with popover (alignment and focusing) and documentation fixes, also removed repetitive code and changed name of 'iconClass'. Missing icons added to common-icons.ts as they weren't displaying.

donna-marijne commented on src/components/icon/iconsets/common-icons.ts:
```diff
@@ -189,6 +189,7 @@ export const commonIcons = [
     'information',
     'inherit-filled',
     'inherit',
+    'input-to-process',
```
I forgot this file exists too, @ashley-hunter shouldn't this be generated during the build?

TaraTodd123 replied:
Ashley left a comment of the ticket about this https://portal.digitalsafe.net/browse/EL-3593 and opened a pull request to fix it.

TaraTodd123 commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.html:
```diff
@@ -20,28 +21,22 @@
 
-        <p>To use ux-icon, ensure that IconModule has been imported, and bind the icon name to name:</p>
+        <p>To use <code>ux-icon</code>, ensure that <code>IconModule</code> has been imported, and bind the
+            icon name to name:</p>
```
Fixed up now and thanks.

TaraTodd123 commented on docs/app/pages/css/css-sections/icons/ux-icons/ux-icons.component.html:
```diff
@@ -1,7 +1,8 @@
     with one of the following classes:</p>
 
-<p  *ngIf="iconClass === 'ux-icon'">To use an icon apply the <code>ux-icon</code><a fragment="icon"
+<p  *ngIf="iconSetClass === 'ux-icon'">To use an icon apply the <code>ux-icon</code><a fragment="icon"
```
Added in now and thanks again

TaraTodd123 commented on docs/app/pages/css/css-sections/icons/ux-icons/icon-preview/icon-preview.component.less:
```diff
@@ -0,0 +1,14 @@
+@import '../../../../../../assets/styles/variables.less';
```
removed
